\documentclass[11pt]{report}

\title{Introduction to the Nextflow pipeline \textbf{CoComics}: Conditional Couplings in (transcript-)omics data.}
\author{Abel Jansma}
\date{\today}
 
 
\begin{document}
\maketitle

\section{Introduction}
This pipeline takes in single cell RNA-seq count matrices, and estimates interactions between the expressed genes. Once nextflow is installed on your machine, the pipeline can be pulled from Github, and runs in a Docker/Singularity container on the data you provide it with. It can run on your local machine, or on a cluster with the Sun Grid Engine scheduling system (like Eddie). 


\section{Requirements}
Nextflow needs to be installed on the machine you want to run the pipeline on, which is as simple as running:

\begin{verbatim}
	 curl -s https://get.nextflow.io | bash
\end{verbatim}

Secondly, you must have access to either Docker or Singularity. Most clusters use Singularity for security reasons, but both will automatically pull the right container from DockerHub. (Conda environments are supported, but not recommended or guaranteed to work.)

\section{Input files}



\section{Parameters to set}

Two classes, to be set in JSON file: \\


algorithmic parameters
\begin{itemize}
	\item 
\end{itemize}

Computational parameters
\begin{itemize}
	\item 
\end{itemize}


\end{document}
